src/main/scala/com/example/solution/services/Reservations.scala

```diff
@@ -0,0 +1,134 @@
+package com.example.solution.services
+
+import cats.data.{NonEmptyList, OptionT}
+import cats.effect._
+import cats.syntax.all._
+import com.example.solution.domain.Ticket
+import com.example.solution.domain.user.{User, UserName, UserSurname}
+import com.example.solution.dto.reservation.{MadeReservation, PickedSeat, ReservationData}
+import com.example.solution.dto.screening.{AvailableSeat, RoomData}
+import skunk._
+import skunk.codec.all._
+import skunk.implicits._
+
+import java.time.LocalDateTime
+import java.time.temporal.ChronoUnit
+
+trait Reservations[F[_]] {
+
+  def makeReservation(data: ReservationData): F[Option[MadeReservation]]
+
+
+}
+
+object LiveReservations {
+
+  def make[F[_] : Sync](
+                         sessionPool: Resource[F, Session[F]],
+                         screenings: Screenings[F]
+                       ): F[Reservations[F]] =
+    Sync[F].delay(
+      new LiveReservations[F](sessionPool, screenings)
+    )
+}
+
+final class LiveReservations[F[_] : BracketThrow : Sync] private(
+                                                                  sessionPool: Resource[F, Session[F]],
+                                                                  screenings: Screenings[F]
+                                                                ) extends Reservations[F] {
+
+  import ReservationQueries._
+
+  private def areSeatsValid(picked: List[PickedSeat],
+                            available: List[AvailableSeat],
+                            seatsPerRow: Int): Boolean = {
+    val isSeatAlreadyTaken = picked.exists(ps => !available.contains(AvailableSeat(ps.row, ps.seatInRow)))
+    val leftSeats = available diff picked.map(ps => AvailableSeat(ps.row, ps.seatInRow))
+    val leftOneEmptySeat = leftSeats.groupBy(_.row).exists(tuple => {
+      val list = tuple._2
+      list.exists(seat => {
+        val seatInRow = seat.seatInRow
+        (seatInRow > 1 && seatInRow < seatsPerRow) &&
+          (!list.exists(as => as.seatInRow == seatInRow - 1) && !list.exists(as => as.seatInRow == seatInRow + 1))
```

Comment:
I believe the logic is written well, because it's rather hard to read.

Reply:
That's true, I guess I could have created couple of additional functions, like isSeatOnTheEdge(seat : Int) : Boolean and seatOnTheLeftTaken : Boolean(same for right). I also think performance of the algorithm is not the best, but the readability would probably be even worse then.